### PR TITLE
Closes #25486: fix sign in button

### LINKED ISSUE #25486: App menu shows "Sign in to sync" menu item until I tap it, even though I'm already signed in to sync
### Steps to reproduce

1. Log into your FxA Sync in Fenix.
2. Kill Fenix.
3. Launch Fenix.
4. Open the app menu.
5. See "Sign in to sync". I think this is a bug!
6. Tap the "Settings" menu item.
7. See at the top of the Settings screen that you are already signed in.
8. Exit the Settings screen.
9. Open the app menu again.
10. See that the "Sign in to sync" menu item has been replaced with your sync account name. This is the result I expected in step 5 above.

Maybe I am experiencing https://github.com/mozilla-mobile/fenix/issues/24112? But I don't think my account is in a bad authentication state.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20628)

### PR DESCRIPTION
**Pull request by mavduevskiy**
### What's wrong with the signin button?

Speaking plainly, the settings menu displays items that are created during the `init` phase of building the layout. The account manager needs some time to get updated, so during the `init` phase of the layout it's not yet ready. We used to show `sign in` even for already singed in users, what looked like a bug. Currently, we [hide](https://github.com/mozilla-mobile/fenix/pull/26494) the `sign in` element completely not to cause confusion.

### What we have tried?

We could observe the account manager and when it is ready update the items. But it would [close](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/menu/src/main/java/mozilla/components/browser/menu/view/MenuButton.kt#102) an already opened menu. There were a few [attempts](https://github.com/mozilla-mobile/fenix/pull/26494#issuecomment-1219870905) [attemps](https://github.com/mozilla-mobile/fenix/pull/21488/files) to go this way, but eventually we decided against forcefully closing the already displayed menu.

### How do we solve similar issues?

`SignIn` isn't the only menu items that has a state, we also have `bookmark` and `pin` menu items. Upon interaction, the menu changes their appearance on the next launch while keeping the same set of items. To determine the current state of the item, menu relies no on parameters but on [functions](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/menu/src/main/java/mozilla/components/browser/menu/item/BrowserMenuImageTextCheckboxButton.kt#50), delegating calculation to the outside.

 ### Could we use already existing solutions?

[TwoState](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/browser/menu/src/main/java/mozilla/components/browser/menu/item/TwoStateBrowserMenuImageText.kt) class sounds like it could solve out problem. But it still requires both titles as parameters – we can't know the name of the user account, so I can't use that.

We could delegate calculating the title – use a function instead of a parameter. It requires little code and follows the existing approach. It also allows us to add the third state: "Authentication in progress.." that was discussed in multiple ([link](https://github.com/mozilla-mobile/fenix/issues/25486#issuecomment-1206411322), [link](https://github.com/mozilla-mobile/fenix/pull/26494#issuecomment-1219096478)) conversations.  

That's the approach I took.

 ### Is there a better solution?
While delegating calculation of the state of a single item to the outside is a solution, it might lead to inconsistency as the state of each individual item is being calculated dynamically. For example, the new items, that are about to be scrolled, might get the updated state, but the items already drawn might have the old state. 

Calculating all of the items based on the current state of the app would be a more consistent way since we don't want to update the already shown menu dynamically. We could add a helper class (aka `provider`) to the builder, that would calculate menu items on each `build`. That would also make `DefaultToolbarMenu` and `HomeMenu` much lighter.

That could be a follow up.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #25486

### CODE REVIEW

**Comment by mavduevskiy** [comment]
@Alexandru2909 @Mugurell @rocketsroger @MatthewTighe 

Pinging you guys because you were involved in the discussion in various issues and PRs related to the issue. Want to get your feedback on the approach I took and the approach I suggested as a follow up. If you want to play around with different ways of solving the issue, take a look at the second commit. Something requires a bit of tinkering on the AC side, pretty basic and obvious though.

**Comment by mavduevskiy** [comment]
probably have to talk with the UX about the "Authentication in progress.." string

**Comment by Mugurell** [comment]
Not sure about adding a new string: "Syncing.." to use until we know the current FXA state.
Seeing "Syncing.." before the option to "Sync and save data" could leave some users wondering about what is syncing and why.

Recording of the app from a Samsung S7 with Android 8:

https://user-images.githubusercontent.com/11428869/211287090-aee53a1c-48e3-4b3b-802b-f8ceb77b286c.mp4

**Comment by Mugurell** [comment]
Regarding the issue it's important to note that it affects only the cold start of the app hence the smaller the fix the better.

If the current scenario of not showing the fxa button if we know about the fxa status is not acceptable I would suggest caching the current fxa state such that it will be available even before the account manager is available.

With `accountManager` being an instance not yet available for our usecase it seems like we have to handle the caching.
Fortunately we already know if a user is logged in or not with the help of
https://github.com/mozilla-mobile/fenix/blob/d077210eb810f4db336ffd8025e5c5dd91cbf6e7/app/src/main/java/org/mozilla/fenix/utils/Settings.kt#L1387

Not sure on whether 
- we should also cache in SharedPreferences the last signed in fxa account **email** to show in the menu (from a security standpoint, might need to check with Christian) or 
- if we know the user is signed in but the `accountManager` is not available show the "Syncing.." text (might need to sync with UX for this new state shown in the menu).

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 1]
```diff
@@ -97,25 +98,19 @@ class HomeMenu(
         }
     }
 
-    private fun syncSignInMenuItem(): BrowserMenuImageText? {
-        val syncItemTitle =
-            if (context.components.backgroundServices.accountManagerAvailableQueue.isReady()) {
-                accountManager.accountProfileEmail ?: context.getString(R.string.sync_menu_sync_and_save_data)
-            } else {
-                null
-            }
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:106]: Looks to me like this is the only difference - show "Syncing.." instead of not showing a menu option at all?
In which case the new text could be added here instead of `null` without the need of a new `BrowserMenuSignIn` class.

Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:106]: If we add text here, then the item will be build with the "syncing..." text and won't be change on the consecutive menu calls, until the user leaves the screen and the item gets recreated. 

The main difference is that `BrowserMenuSignIn` uses a function to get text every time instead of receiving it as a parameter. Calling the function during the `bind`allows `BrowserMenuSignIn` to get the most recent state from the account manager.

I will discuss the phrasing with UX

**Comment by mavduevskiy** [comment]
To sum up, the are two problems in [#25486](https://github.com/mozilla-mobile/fenix/issues/25486):

- If the information is not ready/not available, we show incorrect state and confuse the user.
- If there is no connection at launch, the button always will be in `please, sign-in` state. 

Both problems would be solved if we had access to profile cache, but we, apart from a 2min time-frame, [don't](https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41).

This patch is solving the first issue. The second requires talking to the A-S team, which I will initiate shortly. I also have an idea to update the profile info when `syncNow` succeeds, it would allow the user to see the profile info when the device goes back online. But it's more of a hack and just partially solves the needed cache problem.

**Comment by MatthewTighe** [comment]
> Both problems would be solved if we had access to profile cache, but we, apart from a 2min time-frame, [don't](https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41).

I think @Mugurell's strategy suggested above:
> if we know the user is signed in but the accountManager is not available show the "Syncing.." text (might need to sync with UX for this new state shown in the menu).

would be an interesting and easy one if it gets Product and UX signoff and would allow us to keep the patch relatively small as well as avoid requiring new functionality added to A-S.

**Comment by mavduevskiy** [comment]
@MatthewTighe 
Yes, that's actually what the patch is doing🙂 The UX approved the [Account info](https://github.com/mozilla-mobile/fenix/issues/25486#issuecomment-1377810186) string.

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -165,6 +165,8 @@
     <string name="browser_menu_stop">Stop</string>
     <!-- Browser menu button that opens the addon manager -->
     <string name="browser_menu_add_ons">Add-ons</string>
+    <!-- Browser menu button that opens account settings or -->
+    <string name="browser_menu_account_settings">Account info</string>
```
Comment (gabrielluong) [inline app/src/main/res/values/strings.xml:169]: You might want to consider creating a PR to preland the string since string freeze is tomorrow if you're expecting to  ship this fix for the upcoming release.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+
+internal const val NO_ID = -1
+
+internal fun TextView.setColorResource(@ColorRes textColorResource: Int) {
+    if (textColorResource != NO_ID) {
+        setTextColor(ContextCompat.getColor(context, textColorResource))
+    }
+}
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:24]: Looking at the code, it didn't appear we would ever hit the `NO_ID` case since we are always passing in the primary color. Do we need this?

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (MatthewTighe) [reply app/src/main/res/values/strings.xml:169]: +1 to this suggestion. Looking over this in-depth now, but I expect we will have at least one review cycle so it would be a safe bet to preland the string.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            // without connection to the Internet, but profile can not. So we have to make
+            // sure that we actually have the email in order to show it.
+            email ?: resources.getString(R.string.browser_menu_account_settings)
+        } else {
+            getString(R.string.sync_menu_sync_and_save_data)
+        }
+    } else {
+        resources.getString(R.string.browser_menu_account_settings)
+    }
+}
```
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:52]: Some suggestions for changes:

Since we already have access to the components, we could use the `SyncStore` instead of the `accountManager`. I believe that moving forward we would prefer the `SyncStore` abstraction, both because it follows the `State` library pattern more closely and because it will allow us to begin refactoring some of the complexity in the underlying Sync components. This could be done in combination with the `Settings.signedInFxaAccount` property, something like
```
if (settings.signedInFxaAccount) {
    syncStore.state.account?.email ?: R.string.browser_menu_account_settings
} else {
    R.string.sync_menu_sync_and_save_data
}
```

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ */
+
+class BrowserMenuSignIn(
+    private val label: () -> String,
+    @DrawableRes
+    imageResource: Int,
+    @ColorRes
+    private val textColorResource: Int,
+    listener: () -> Unit = {},
+) : BrowserMenuImageText(
```
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:71]: Personally, I think I would prefer to use an inline anonymous object and move the logic from the extension function above into `DefaultToolbarMenu`. I think the likelihood that this class is re-used is pretty low. That said, my opinion is not that strong, so keep it if you want to.

I have illustrated this suggestion in more depth in the context of the other file.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt [hunk 6]
```diff
@@ -349,27 +350,15 @@ open class DefaultToolbarMenu(
...
-            }
+    private fun syncMenuItem(): BrowserMenuItem {
+        return BrowserMenuSignIn(
+            label = { context.getSignInLabel() },
+            R.drawable.ic_signed_out,
+            primaryTextColor(),
+        ) {
+            onItemTapped.invoke(
+                ToolbarMenu.Item.SyncAccount(accountManager.accountState),
+            )
```
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/DefaultToolbarMenu.kt:361]: Using an anonymous object instead here would look like:

```
private fun syncMenuItem() = object : BrowserMenuImageText(
    label = syncMenuItemText()
    ...
) {
    override fun bind() {
        super.bind(menu, view)
        val textView = view.findViewById<TextView>(R.id.text)
        textView.text = syncMenuItemText()
        textView.setColorResource(textColorResource)
     }
}

private fun syncMenuItemText(): String {
  // see above suggestion
}
```

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ * @param textColorResource Optional ID of color resource to tint the text.
+ * @param listener Callback to be invoked when this menu item is clicked.
+ */
+
+class BrowserMenuSignIn(
+    private val label: () -> String,
+    @DrawableRes
+    imageResource: Int,
+    @ColorRes
+    private val textColorResource: Int,
```
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:69]: nit: if you keep the class, the annotations could be moved onto the same line as the parameter name for better readability

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt (hunk 4 above)
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:52]: Ah, I see now this function is used in both the `DefaultToolbarMenu` and the `HomeMenu`. In that case, we clearly need to keep it. You could consider moving it into `ext/SyncState` instead, as that feels like a more common location between `home` and `components/toolbar` to me. Something like:

```
SyncState.getSignInLabel(settings: Settings) String {
    ...
}
```

This would have the added benefit of being more easily testable than an extension defined on `Context`.

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (mavduevskiy) [reply app/src/main/res/values/strings.xml:169]: Good suggestion! I wish I didn't miss it yesterday.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt (hunk 3 above)
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:24]: I copied it from the parent class, to follow the same parameter validation rules. I am not sure we need it, since we are the only users of the class. Removing.

**Comment by mavduevskiy** [comment]
@MatthewTighe 
Thanks, Matt, for the suggestions! Maybe keeping the extra class but changing an extension function into private would solve the issue? It could be reused and easily tested. Added the tests, thanks for the reminder.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+ * A menu item for displaying account information. The item calculates the label on every bind call,
+ * to provide each menu with the latest account manager state.
+ *
+ * @param label The function that provides the visible label of this menu item.
+ * @param imageResource ID of a drawable resource to be shown as icon.
+ * @param textColorResource Optional ID of color resource to tint the text.
+ * @param listener Callback to be invoked when this menu item is clicked.
+ */
+
+class BrowserMenuSignIn(
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:29]: I am curious about the rationale about this class since it doesn't look like it is doing anything extra that `BrowserMenuImageText` couldn't. What's strange to me was that we are passing a `label` param which will effectively not be utilized when we also just bind by the text view with `getLabel` anyways. So, I wouldn't expose `label` either as a param and just add the default `String()` value in line 35 if this class was needed.

It looks possible to me that we could just have the `getLabel` (maybe rename to `getSignInLabel`) function in `DefaultToolbarMenu.kt` and call it via

```
BrowserMenuImageText(
                    getSignInLabel(),
                    R.drawable.ic_signed_out,
                    primaryTextColor(),
                ) {
                    onItemTapped.invoke(
                        ToolbarMenu.Item.SyncAccount(accountManager.accountState),
                    )
                }
```

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+     */
+    @VisibleForTesting
+    internal fun getLabel(context: Context): String {
+        return with(context) {
+            val isAuthenticated = components.settings.signedInFxaAccount
+            val isAccountManagerReady = components.backgroundServices.accountManagerAvailableQueue.isReady()
+            val email = components.backgroundServices.syncStore.state.account?.email
+
+            if (isAccountManagerReady) {
+                if (isAuthenticated) {
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:63]: @Mugurell I remember discussing with you about adding `enum` for the FXA sign in states that are available so we wouldn't have to do all this if else checks and can easily encapsulate the states in a class. Do you happen to remember where that is and if we could reuse that here?

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt (hunk 8 above)
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:29]: My though process is described in the body of the PR. Also, it looks like Petru asked the same [question](https://github.com/mozilla-mobile/fenix/pull/28409#discussion_r1064500811).

The version that you suggested will calculate the `label` once in the constructor, and apply the text every `bind` call. In contrast, `BrowserMenuSignIn` will calculate it on every bind call (to support slow internet or opening the menu while `accountManager` is still busy).

You are right about passing the label parameter, it's a bit I forgot to remove. Was trying to write a test that would make sure that the value from constructor gets `overriden`, but got stuck trying to mock `onBind()` method.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt (hunk 9 above)
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: Good memory!
Yeah, we have a [FenixAccountManager](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/components/accounts/FenixAccountManager.kt) which already contains the `accountState` and `accountProfileEmail` info.
While the `FxaAccountManager` that is wrapped in `FenixAccountManager` has a lengthy initialization it seems like it is called in `FenixApplication` also so it should be ready by the time the menu is opened.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: @mavduevskiy Could you look at using `FenixAccountManager`? This should also help simplify the lengthy comment you have here because the account state will already be documented (unless the freshness state is not accounted there then we should just make sure that state is also documented).

Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: @gabrielluong 
One of the main goals is to be able to accommodate the `The account manager needs some time to get updated` case, but `FenixAccountManager` reports `no_account` state in that case. `no_account` and `no_information_yet` are different UI cases: we ask to sign in in the first case, and show general "account info" in the second. 
As for the `profile` information and `is_signed_in` state I went with @MatthewTighe [suggestion](https://github.com/mozilla-mobile/fenix/pull/28409#discussion_r1067545553).

About the comment, I didn't find any documentation or logic to accommodate the state when we have an account but not a profile, hence the comment on why in the `isAuthenticated` we still want to check if the profile is not null. Should be able to remove it once the [caching mechanism](https://github.com/mozilla/application-services/pull/5333) gets fully available.

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: > While the `FxaAccountManager` that is wrapped in `FenixAccountManager` has a lengthy initialization it seems like it is called in `FenixApplication` also so it should be ready by the time the menu is opened.

We introduced the SyncStore that holds the account state (see: [SyncStatus](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/store/SyncState.kt)) as well to have one place where you can know this information that also follows our store pattern.

The hope is that we can remove the FenixAccountManager and other duplications as well so we can just query with `syncStore.state.status`.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: It sounds like we have another approach to investigate with the `SyncStore`. I don't have the necessary bandwidth to look into this too deeply right now, but it would be good to know what you decide on with all this new information.

Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: This is well beyond the scope of this PR at this point, but I wanted to address my thoughts in context:

If the enums defined in `FenixAccountManager` are a desirable method of determining the state of account state then I think it would be preferable to add them to the `SyncStore`, in order to continue simplifying Sync code. Looking `FenixAccountManager`, I don't see any reason why we wouldn't want something similar in the `SyncStore`. We could even investigate an update to `SyncState` such that we had something like:

```
data class SyncState(
    val accountState: AccountState? = null,
) : State

sealed class AccountState {
    data class Authenticated(val account: Account) : AccountState()
    object NoAccount : AccountState()
    object NeedsReauth : AccountState()
}
```

which seems better anyway to me than needing intrinsic knowledge about `SyncState.account`.

> What I am implicitly looking for is some form of simplification similar to FenixAccountManager to all of this if else checks to get the account state and have it broadly documented and be reusable since this has been a common problem and these checks are still spread out in various parts of the code.

The above I think would address this as well as we are able, but does not deal with one of the underlying issues - the only mechanism we have for caching whether users are authenticated is through `Settings`. I also think it would be worth adding that cache to A-C, but I am not sure what that would look like

Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:62]: In general, the `SyncStore` is meant to address the fundamental problem that @mavduevskiy has described in the body of the PR. There are a couple options listed there, but another is to use reactive programming. That is the intent of the observable Stores, but I agree with @mavduevskiy that overriding the `bind` function here is a good choice because it is a smaller lift than making starting to make this Toolbar use reactive programming. If this were to be updated to Compose I think that would be a good time to make a more fundamental change like that.

**Comment by MozillaNoah** [comment]
Based on the level of discussion here, it feels like we need to take a new look at `BrowserMenu` in the AC layer. I can't help but think this all could be avoided in the future when we convert our menus, in general, to Compose. We'll be able to keep the modularity of the current menu's design, but it would also make it trivial to respond to state changes. Is this something we should put on our long term tech roadmap?

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+     * occur when the menu is opened before the AccountManager has finished updating its state, or
+     * when there is no connection and AccountManager doesn't return us the profile because its cache
+     * has 2 minutes 'freshness':
+     * https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41
+     */
+    @VisibleForTesting
+    internal fun getLabel(context: Context): String {
+        return with(context) {
+            val isAuthenticated = components.settings.signedInFxaAccount
+            val isAccountManagerReady = components.backgroundServices.accountManagerAvailableQueue.isReady()
```
[the remark this answers is not in the document]
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt]: True! I didn't realize that `components.settings.signedInFxaAccount` is cached and will be ready immediately. Good catch!

File: app/src/test/java/org/mozilla/fenix/menu/BrowserMenuSignInTest.kt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    fun `WHEN accountManager is not ready, THEN show the general account info text`() {
+        every { components.backgroundServices.accountManagerAvailableQueue.isReady() } returns false
+        every { components.backgroundServices.syncStore.state.account } returns null
+        every { components.settings.signedInFxaAccount } returns false
+
+        assertEquals(
+            testContext.getString(org.mozilla.fenix.R.string.browser_menu_account_settings),
+            BrowserMenuSignIn(R.color.black).getLabel(context),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (mavduevskiy) [reply app/src/test/java/org/mozilla/fenix/menu/BrowserMenuSignInTest.kt]: yep, adjusted the tests

**Comment by mavduevskiy** [comment]
@MozillaNoah 
We just discussed that with @jonalmeida today, and he mentioned that converting the existing menu to `Compose` might take up to 6 months. I wonder if there are any particular parts of functionality that make the `Compose` implementation that hard of task.

**Comment by jonalmeida** [comment]
> he mentioned that converting the existing menu to `Compose` might take up to 6 months.

Sorry, I was being a bit facetious when I said that. It will take a long time and we should better plan and construct out what that menu API would look like.

I say menu API because it's something we need to build deliberately that will be a component for Fenix and Focus.

**Review by MatthewTighe — APPROVED** [review]
I have a couple small nits left, but I otherwise feel you have addressed my concerns and that this LGTM. 

Since this has had such active conversation, I would prefer to wait to land it until after tomorrow in hopes that anyone else interested has the chance to additionally approve or request changes.

Also, friendly reminder to squash before landing 🙂

**Comment by mavduevskiy** [comment]
@MatthewTighe 
Thanks, Matt! I addressed your feedback, pinging others for the second approval 
@Mugurell @jonalmeida @MozillaNoah @gabrielluong

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * A menu item for displaying account information. The item calculates the label on every bind call,
+ * to provide each menu with the latest account manager state.
+ *
+ * @param label The function that provides the visible label of this menu item.
+ * @param imageResource ID of a drawable resource to be shown as icon.
+ * @param textColorResource Optional ID of color resource to tint the text.
+ * @param listener Callback to be invoked when this menu item is clicked.
+ */
+
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:28]: nit: extra new line here

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        textView.setTextColor(ContextCompat.getColor(textView.context, textColorResource))
+    }
+
+    /**
+     * Determines the proper label for the sign in button.  There are three states:
+     * user is signed-in, please sign-in, and we are not sure what's happening. The last state does
+     * occur when the menu is opened before the AccountManager has finished updating its state, or
+     * when there is no connection and AccountManager doesn't return us the profile because its cache
+     * has 2 minutes 'freshness':
+     * https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:52]:
```suggestion
     * Returns the proper label for the sign in menu button.
```
I think the comments could be simplified here since the code are now just doing simple checks for whether or not the user is signed in or not. So, I could reasonably imagine that we don't have to care about the 3 states and especially this 'freshness' state where we are saying we don't know what's happening - it's just kinda concerning to say we aren't sure what's happening in the landed code.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        val isSignedIn = components.settings.signedInFxaAccount
+        val email = components.backgroundServices.syncStore.state.account?.email
+
+        if (isSignedIn) {
+            // It sounds a bit counterintuitive, but singed-in state doesn't mean we
+            // have the profile information. The state can be (and usually is) restored
+            // without connection to the Internet, but profile cache is accessible just
+            // for two minutes. So we have to make sure that we actually have the email
+            // in order to show it. Improving cache work progress:
+            // https://github.com/mozilla/application-services/issues/5328
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:65]: Could we remove this comment as well? I imagine this is no longer an issue that no longer needs to documented due to our `SyncStore` usage.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import androidx.annotation.DrawableRes
+import androidx.annotation.VisibleForTesting
+import androidx.core.content.ContextCompat
+import mozilla.components.browser.menu.BrowserMenu
+import mozilla.components.browser.menu.item.BrowserMenuImageText
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+
+/**
+ * A menu item for displaying account information. The item calculates the label on every bind call,
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt:20]:
```suggestion
 * A menu item for displaying account information. The item computes the label on every bind call,
```

File: app/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -165,6 +165,8 @@
     <string name="browser_menu_stop">Stop</string>
     <!-- Browser menu button that opens the addon manager -->
     <string name="browser_menu_add_ons">Add-ons</string>
+    <!-- Browser menu button that opens account settings or -->
```
Comment (gabrielluong) [inline app/src/main/res/values/strings.xml:168]:
```suggestion
    <!-- Browser menu button that opens account settings -->
```

File: app/src/test/java/org/mozilla/fenix/menu/BrowserMenuSignInTest.kt [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    @Test
+    fun `WHEN signed in and has profile data, THEN show email`() {
+        every { components.backgroundServices.syncStore.state.account } returns account
+        every { components.settings.signedInFxaAccount } returns true
+
+        assertEquals(account.email, BrowserMenuSignIn(R.color.black).getLabel(context))
+    }
+
+    @Test
+    fun `WHEN not signed in, THEN show the please-sign-in text`() {
```
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/menu/BrowserMenuSignInTest.kt:48]:
```suggestion
    fun `WHEN not signed in, THEN show the sync and save data text`() {
```
nit: I wanted us to be a bit more precise about what the actual text is being displayed to not confuse anyone when the text doesn't match up. `R.string.sync_menu_sync_and_save_data` maps to "Sync and save data".
This is the same for the test comments below.

**Comment by gabrielluong** [comment]
I think the code looks good to me, but I did not go into any depth to test this or look into `SyncStore`. I was mainly focus on being precise and correct with our comments since we have already drastically simplified most of the edge cases concern with our new approach. No further re-review needed from me.

**Comment by mavduevskiy** [comment]
@jonalmeida 
Thanks, Jonathan!😅
I fixed @gabrielluong nits, they are actually very helpful, thanks, gl!

File: app/src/test/java/org/mozilla/fenix/menu/BrowserMenuSignInTest.kt [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        every { components.backgroundServices.syncStore.state.account } returns null
+
+        assertEquals(
+            testContext.getString(org.mozilla.fenix.R.string.sync_menu_sync_and_save_data),
+            BrowserMenuSignIn(R.color.black).getLabel(context),
+        )
+    }
+
+    @Test
+    fun `WHEN not signed in and has profile data, THEN show the please-sign-in text`() {
```
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/menu/BrowserMenuSignInTest.kt:59]:
```suggestion
    fun `WHEN not signed in and has profile data, THEN show the sync and save data text`() {
```
